### PR TITLE
Support for GNOME 3.26 - 3.38, and GNOME 40

### PR DESCRIPTION
**Pull request by mwilck**
After waiting for a long time, I'm submitting my GNOME shell compatibility work for argos here. See #108 and #111.

### CODE REVIEW

**Comment by real-or-random** [comment]
I've been running this branch on 40 for a while and I can confirm it works fine.

**Comment by dvogel** [comment]
I've been using this with gnome-shell 3.36.7-0ubuntu0.20.04.1 for a few weeks and I haven't encountered any issues.

**Comment by mwilck** [comment]
@dvogel and @real-or-random, thanks for the feedback. Please note that I added the shellVersion patch on top just today, so if you could update to exactly this version I'd be grateful.

**Comment by dvogel** [comment]
> @dvogel and @real-or-random, thanks for the feedback. Please note that I added the shellVersion patch on top just today, so if you could update to exactly this version I'd be grateful.

I had not noticed that. I just updated to this latest commit and restarted gnome-shell. No issues so far.

**Comment by logix2** [comment]
I'm using this too with GNOME Shell 40.1 with no issues.

File: argos@pew.worldwidemann.com/menuitem.js [hunk 1]
```diff
@@ -13,33 +13,32 @@ const Lang = imports.lang;
...
-
-    super({
-      activate: hasAction,
-      hover: hasAction,
-      can_focus: hasAction
-    });
+// "constructor" aka "init function" for ArgosMenuItem.
+// Defined as ordinary function, referencing "this"; these references will
+// be resolved by later bind() calls.
+// Utilities.MakeSimpleClass() is used below to actually define the class.
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/menuitem.js:25]: Ughh... but I guess this is the reality if you want code to work across multiple GNOME Shell versions :frowning_face:

File: argos@pew.worldwidemann.com/metadata.json [hunk 2]
```diff
@@ -3,6 +3,14 @@
   "name": "Argos",
   "description": "Create GNOME Shell extensions in seconds",
   "url": "https://github.com/p-e-w/argos",
-  "version": 3,
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/metadata.json:6]: Why was this removed?

File: argos@pew.worldwidemann.com/utilities.js [hunk 3]
```diff
@@ -167,7 +172,8 @@ function parseLine(lineString) {
   }
 
   line.hasAction = line.hasOwnProperty("bash") || line.hasOwnProperty("href") ||
-    line.hasOwnProperty("eval") || line.refresh === "true";
+	line.hasOwnProperty("eval") || 
+	(line.hasOwnProperty("refresh") && line.refresh === "true");
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:176]: No tabs please.

File: argos@pew.worldwidemann.com/utilities.js [hunk 4]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
   });
 }
 
+function getShellVersion(str) {
+
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:282]: Delete this empty line.

File: argos@pew.worldwidemann.com/utilities.js [hunk 5]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+    let v = str.split(".");
+    let n = 0;
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:284]: Please use verbose variable names that are self-documenting.

File: argos@pew.worldwidemann.com/utilities.js [hunk 6]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+
+	if (major >= 40)
+	    n = 30000 + major * 100 + minor;
+
+    } else if (v.length == 3 && v[0] == "3") {
+	n = v.map(Number).reduce(
+	    function(a, x) {
+		return 100 * a + x;
+	    });
+
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:313]: Delete this empty line.

File: argos@pew.worldwidemann.com/utilities.js [hunk 7]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
     let [line] = source.read_line_finish(result);
 
     if (line === null) {
       callback(null);
     } else {
-      callback(imports.byteArray.toString(line) + "\n");
+      if (shellVersion <= 33400)
+	callback(String(line) + "\n");
+      else
+	callback(imports.byteArray.toString(line) + "\n");
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:337]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 8]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+	    });
+
+    };
+
+    if (n == 0) {
+	log("argos: Unsupported GNOME shell version " + str);
+	return 0;
+    }
+
+    // log("argos: GNOME shell version " + str + " => " + n);
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:321]: Remove this.

File: argos@pew.worldwidemann.com/utilities.js [hunk 9]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+
+function makeSimpleClass(BaseClass, getSuperArgs, initFn, name) {
+  if (shellVersion < 33200) {
+    return new Lang.Class({
+      Name: name,
+      Extends: BaseClass,
+      _init: function(...args) {
+	this.parent(getSuperArgs(...args));
+	initFn.bind(this)(...args);
+      }
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:358]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 10]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+    return class extends BaseClass {
+      constructor(...args) {
+	super(getSuperArgs(...args));
+	initFn.bind(this)(...args);
+      }
+    }
+  } else {
+    return GObject.registerClass(
+      {
+	GTypeName: name
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:370]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 11]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+	this.parent(getSuperArgs(...args));
+	initFn.bind(this)(...args);
+      }
+    });
+  } else if (shellVersion < 33400) {
+    return class extends BaseClass {
+      constructor(...args) {
+	super(getSuperArgs(...args));
+	initFn.bind(this)(...args);
+      }
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:365]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 12]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+  } else {
+    return GObject.registerClass(
+      {
+	GTypeName: name
+      },
+      class extends BaseClass {
+	_init(...args) {
+	  super._init(getSuperArgs(...args));
+	  initFn.bind(this)(...args);
+	}
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:376]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 13]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+      });
+  }
+}
+
+if (shellVersion <= 33400)
+    var AltSwitcher = imports.ui.status.system.AltSwitcher;
+else
+  var AltSwitcher = GObject.registerClass(
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:384]: Put this in a `{}` block, it's too big for inline.

File: argos@pew.worldwidemann.com/utilities.js [hunk 14]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+	_init(...args) {
+	  super._init(getSuperArgs(...args));
+	  initFn.bind(this)(...args);
+	}
+      });
+  }
+}
+
+if (shellVersion <= 33400)
+    var AltSwitcher = imports.ui.status.system.AltSwitcher;
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:382]: Indentation doesn't match the `else` clause.

File: argos@pew.worldwidemann.com/utilities.js [hunk 15]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+          }
+          childToShow.add_action(this._clickAction);
+
+          let hasFocus = this.contains(global.stage.get_key_focus());
+          this.set_child(childToShow);
+          if (hasFocus)
+            childToShow.grab_key_focus();
+
+	  // The actors might respond to hover, so
+          // sync the pointer to make sure they update.
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:454]: Indentation.

File: argos@pew.worldwidemann.com/utilities.js [hunk 16]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+
+if (shellVersion <= 33400)
+    var AltSwitcher = imports.ui.status.system.AltSwitcher;
+else
+  var AltSwitcher = GObject.registerClass(
+    class AltSwitcher extends St.Bin {
+      _init(standard, alternate) {
+        super._init();
+        this._standard = standard;
+        this._standard.connect('notify::visible', this._sync.bind(this));
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:389]: Please use double quotes like elsewhere in the code.

File: argos@pew.worldwidemann.com/utilities.js [hunk 17]
```diff
@@ -272,15 +278,211 @@ function spawnWithCallback(workingDirectory, argv, envp, flags, childSetup, call
...
+	}
+      });
+  }
+}
+
+if (shellVersion <= 33400)
+    var AltSwitcher = imports.ui.status.system.AltSwitcher;
+else
+  var AltSwitcher = GObject.registerClass(
+    class AltSwitcher extends St.Bin {
```
Comment (p-e-w) [inline argos@pew.worldwidemann.com/utilities.js:385]: I'm guessing this is copied/adapted from the code that used to be in GNOME Shell? If so, this should be documented.

I also don't get why we fall back to `imports.ui.status.system.AltSwitcher` on older Shell versions. If the entire code must be copied anyway, we might as well use it always, no?

**Review by p-e-w — CHANGES_REQUESTED** [review]
Thank you very, very much for this!

Style nits only, then this is ready to go.

**Comment by insign** [comment]
Thank you so so so much

**Comment by mwilck** [comment]
@p-e-w, @insign, thanks a lot. 

Anyone reading this, I could use help with these style fixes. As mentioned before, I'm really not an experienced Javascript programmer *at all*. I really just did what was necessary to make argos work again. I'll look into the requested changes/improvements, but they are quite many, and it'll take time. So if anyone has time to help out (ideally in the form of PRs against my master branch), I'd greatly appreciate it and we might be able to achieve a mainline merge more quickly.

**Comment by insign** [comment]
@mwilck @p-e-w using eslint we can check the styles before push.

**Comment by mwilck** [comment]
@insign, I believe you but my experience with this tool is exactly zero.

**Comment by insign** [comment]
@mwilck it's not you. @p-e-w needs to define a standard to we follow. PEW, can you do this?

**Comment by p-e-w** [comment]
The standard is "what the rest of the code looks like". I'm not going to write up formal requirements, sorry.

I don't understand what the problem is here. I'm asking that tabs be changed to spaces (which all of the existing code uses), single quotes be changed to double quotes (which all of the existing code uses), indentation matches the surrounding code, etc.. That has nothing to do with JavaScript per se. I'm not asking for any functional change, just that the style of the new code fits with the code that is already there so that the quality of the codebase doesn't degrade.

**Comment by mwilck** [comment]
Understood. I haven't said that it's a problem, just that I need to find time for making the changes.

**Comment by marianrh** [comment]
@p-e-w and @mwilck, thank you both for Argos and for creating this branch! I hope it can merged soon so we can enjoy Argos in Ubuntu 21.10 with GNOME 40 :)

File: argos@pew.worldwidemann.com/utilities.js (hunk 4 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:309]: The empty line this comment originally addressed (after `function getShellVersion(str) {`) has been removed.

File: argos@pew.worldwidemann.com/utilities.js (hunk 9 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:355]: Fixed

File: argos@pew.worldwidemann.com/utilities.js (hunk 11 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:362]: Fixed

File: argos@pew.worldwidemann.com/utilities.js (hunk 13 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:384]: Done

File: argos@pew.worldwidemann.com/utilities.js (hunk 15 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:453]: Fixed

File: argos@pew.worldwidemann.com/utilities.js (hunk 17 above)
Comment (jefferyto) [reply argos@pew.worldwidemann.com/utilities.js:385]: > I'm guessing this is copied/adapted from the code that used to be in GNOME Shell? If so, this should be documented.

I have added a comment - let me know if it is not sufficient.

**Comment by jefferyto** [comment]
@p-e-w would appreciate it if you can take another look at this PR, I have addressed most (if not all) of your style concerns :pray:

**Comment by kinduff** [comment]
@p-e-w Would love to see this merged, either by you, or adding @mwilck as a maintainer for the great support it has provided. I'm sure I'm not the only one willing to donate in order to get 40-41 supported.

File: argos@pew.worldwidemann.com/metadata.json (hunk 2 above)
Comment (p-e-w) [reply argos@pew.worldwidemann.com/metadata.json:6]: It's not clear what the best course of action is here. Should `version` be set in the `metadata.json` that is uploaded to extensions.gnome.org, or does extensions.gnome.org do so itself? It's reassuring to see that the extensions system remains an underdocumented mess.

**Comment by p-e-w** [comment]
**MERGED!!!**

I would like to thank @mwilck, @jefferyto, @michel-slm, and @rammie for their combined effort to finally bring Argos into GNOME's present. While I don't use Argos myself anymore, it feels great to see that others care about the project enough to invest time to keep it working.

**Pull request merged**